src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+            d: d as u64,
+            q: 2.0 * lbar_over_d,
+            b: ((w * lbar_over_d + 1.0) * exponential * d * u_f64 + d).floor() as u64,
+        }
+    }
+
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
+        let sec_param = params.lambda_rel.max(params.lambda_sec).ceil() as u64;
+        let n_p_f64 = params.n_p as f64;
+        let n_f_f64 = params.n_f as f64;
+        let loge = f64::from(LOG2_E);
+
+        let u_f64 = {
+            let numerator = params.lambda_sec + params.lambda_rel.log2() + 5.0 - loge.log2();
+            let denominator = (n_p_f64 / n_f_f64).log2();
+            (numerator / denominator).ceil()
+        };
+
+        let ratio = 9.0 * n_p_f64 * loge / ((17.0 * u_f64).powi(2));
+        let s1 = ratio - 7.0;
+        let s2 = ratio - 2.0;
+
+        if s1 < 1.0 || s2 < 1.0 {
+            // Small case, i.e. n_p <= λ^2
+            Self::param_small_case(params, u_f64, sec_param)
+        } else {
+            let lambda_rel2 = params.lambda_rel.min(s2);
+            if u_f64 < lambda_rel2 {
+                // Case 3, Theorem 14, i.e.  n_p >= λ^3
+                Self::param_high_case(params, u_f64, lambda_rel2, sec_param)
+            } else {
+                // Case 2, Theorem 13, i.e. λ^3 > n_p > λ^2
+                Self::param_mid_case(params, u_f64, s1, sec_param)
+            }
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    // Round candidate tuple
+    s_list: Vec<Element>,
+    /// Round candidate hash
+    h: Hash,
+    /// Round candidate hash mapped to [1, n_p]
+    h_u64: u64,
+    /// Approximate size of set Sp to lower bound
+    n_p: u64,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(
+        first_input: &[u8],
+        second_input: &[u8],
+        n_p: u64,
+        sec_param: u64,
+    ) -> (Hash, Option<u64>) {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H1");
+        hasher.update(first_input);
+        hasher.update(second_input);
+        let digest: Hash = hasher.finalize().into();
+        (digest, utils::sample_uniform(&digest, n_p, sec_param))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: u64, t: u64, n_p: u64, sec_param: u64) -> Option<Self> {
+        let (h, h_u64_opt) = Self::h1(
+            v.to_be_bytes().as_ref(),
+            t.to_be_bytes().as_ref(),
+            n_p,
+            sec_param,
+        );
+        h_u64_opt.map(|h_u64| Self {
+            v,
+            t,
+            s_list: vec![],
```

Comment:
There is a subtle difference between using `vec![]` and `Vec::new()` - the latter uses no macro expansion while the former expands to something like this :) 
```
macro_rules! vec {
    () => {
        {
            let mut v = Vec::new();
            v
        }
    };
    ($elem:expr; $n:expr) => { /* ... */ };
    ($($x:expr),+ $(,)?) => { /* ... */ };
}
```
Thus, using `Vec::new()` here is slightly more efficient in terms of code to process by the compiler, but the runtime difference is negligible, so primarily a matter of style :)

Reply:
Updated